### DISCUSSION TITLE
Ability to provide warnings

### DISCUSSION THREAD

**Discussion by psychedelicious** [colinhacks/zod#2009] [Q&A] [answered]
I'm very much unsure if this belongs in zod - probably not - but I'm similarly unsure of how to approach this. 

I'd like to validate with zod but also provide warnings or hints to users in certain situations.

```typescript
const MyNumberSchema = z.number().min(0).max(1)
const MyNumberWarningSchema = z.number().min(0).max(0.1)

const MyNumberWithWarning = MyNumberSchema.warn(MyNumberWarningSchema)
```

I can imagine `.warn()` taking a normal zod schema, but handling it differently:
- If a value is *valid* when parsed against the warning schema, it generates a warning
- Warnings are nearly identical to errors in how they work, except `warning` is used instead of `error` in the types and objects generated.

The only way I can think of to do this now is to generate a second "inverse" schema, then change the error messages to be more warning-like. 

Thoughts?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const MyNumberWarningSchema = z.number().min( 0 ).max( 0.1 )
    .transform( () => 'value is less than 0.1' )

console.log( MyNumberWarningSchema.parse( 0.05 ) )
// value is less than 0.1
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by psychedelicious** [comment]
That's a piece of the puzzle and very useful - but it's not quite what I am after.

I think this is a better example of the desired functionality.

```javascript
const TooLowWarningSchema = z
  .number()
  .min(0)
  .max(0.1)
  .transform((arg) => `${arg} is dangerously low`);

const TooHighWarningSchema = z
  .number()
  .min(0.9)
  .max(1)
  .transform((arg) => `${arg} is dangerously high`);

const SpecificValueWarningSchema = z
  .literal(0.05)
  .transform((arg) => `${arg} may cause undesired effects`);

const MyNumberSchema = z.number().min(0).max(1);
.warn([TooLowWarningSchema, TooHighWarningSchema, SpecificValueWarningSchema]);

MyNumberSchema.safeParse(0.05)
// {
//  success: true,
//  warnings: ["0.05 is dangerously low", "0.05 may cause undesired effects"],
//  data: 0.05
// };
  ```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const TooLowWarningSchema = z.number().min( 0 ).max( 0.1 )
    .transform( ( arg ) => `${ arg } is dangerously low` )

const TooHighWarningSchema = z.number().min( 0.9 ).max( 1 )
    .transform( ( arg ) => `${ arg } is dangerously high` )

const SpecificValueWarningSchema = z.literal( 0.05 )
    .transform( ( arg ) => `${ arg } may cause undesired effects` )

const MyNumberSchema = z.number().min( 0 ).max( 1 ).transform( value => {
    const warnings = [
        TooLowWarningSchema,
        TooHighWarningSchema,
        SpecificValueWarningSchema
    ].map( schema => {
        const result = schema.safeParse( value )
        return result.success ? result.data : null
    } ).filter( Boolean )
    return { value, warnings }
} )

console.log( MyNumberSchema.safeParse( 0.05 ) )
// {
//     success: true,
//     data: {
//       value: 0.05,
//       warnings: [ '0.05 is dangerously low', '0.05 may cause undesired effects' ]
//     }
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by psychedelicious** [comment]
That's perfect and gives me some new ways to think about using `zod`. Thanks for your help, Jacob.

**Reply by JacobWeisenburger** [reply]
Thanks for the sponsorship!
